src/cache.ts

```diff
@@ -102,13 +113,14 @@ export class ListWatch<T extends KubernetesObject> implements ObjectCache<T>, In
     private async doneHandler(err: any): Promise<any> {
         this._stop();
         if (err) {
-            this.callbackCache[ERROR].forEach((elt: ObjectCallback<T>) => elt(err));
+            this.callbackCache[ERROR].forEach((elt: ObjectCallback<T>) => elt(undefined, err));
```

Comment:
Hmmm. This makes it a breaking change? It also has the error at the end, which may be unintuitive to folks used to Node's error-first callbacks... But moving the error first would be even more breaking...

Not sure what's the best approach here...

Would be a travesty in TypeScript world to allow `on()` to have overloads to accept different parameter types? i.e. allow `on(verb: string, fn: ObjectCallback<T>): void;` as well as `on(verb: string, fn: ErrorCallback): void;`? While it wouldn't enforce the correct callback type based on the `verb`, it would at least be non-breaking? Not sure if there's a better way to express this in TS...

Reply:
I did also try it that way:

`on(verb: string, fn: ObjectCallback<T> | ErrorCallback | ConnectCallback)`

But I felt like that was even messier, because suddenly the user can accidentally pass a `ConnectCallback` to a `on('add', ...)`

Maybe we could do:

```typescript
export type UntypedCallback = (err?: any, obj?: any) => void;
export type TypedCallback<T extends KubernetesObject> = (err?: any, obj?: T) => void;
export type ObjectCallback<T extends KubernetesObject> = (obj?: T, err?: any) => void;

...

on(verb: string, fn: ObjectCallback | TypedCallback | UntypedCallback)
...
```

wdyt? That has the benefit of not being a breaking change, and also adding in the more node-style callbacks.

Reply:
Although, I'm not actually sure that TS can handle that since those two callback types may be indistinguishable...

Reply:
Would TS handle that at runtime or just compile time?

Did TS ever get enums? If the first param for `on()` was enum, rather than a `string`, then it could have multiple signatures based on enum type... Which would probably still be a breaking change from TS perspective, but possibly the old `string` signature could still be there, just with `any` callback so that people can upgrade their code eventually, even if they would lose out a little bit in terms of potentially accidentally passing the wrong callback?

Also I wonder if passing the wrong callback is even a problem for people using TS, because surely someone must have noticed by now that the callback for the `error` event is not right?

This is a common pattern (`on()` with different callbacks for different event types) - surely someone must have solved this by now? This is used in node itself (in streams and elsewhere) - I wonder how do they define the types 🤔 Let's see... 👀

Reply:
was there a follow-up on this @brendandburns ?

Reply:
Thanks for the pointer! I fixed this using type overloading.